File: UIView+Borders.m
//
//  UIView+Borders.m
//
//  Created by Aaron Ng on 12/28/13.
//  Copyright (c) 2013 Delve. All rights reserved.
//

#import "UIView+Borders.h"


@implementation UIView(Borders)

//////////
// Top
//////////

-(CALayer*)createTopBorderWithHeight: (CGFloat)height andColor:(UIColor*)color{
    return [self createTopBorderWithHeight:height color:color leftOffset:0 rightOffset:0 andTopOffset:0];
}

-(UIView*)createViewBackedTopBorderWithHeight: (CGFloat)height andColor:(UIColor*)color{
    return [self createViewBackedTopBorderWithHeight:height color:color leftOffset:0 rightOffset:0 andTopOffset:0];
}

-(void)addTopBorderWithHeight: (CGFloat)height andColor:(UIColor*)color{
    [self addTopBorderWithHeight:height color:color leftOffset:0 rightOffset:0 andTopOffset:0];
}

-(void)addViewBackedTopBorderWithHeight: (CGFloat)height andColor:(UIColor*)color{
    [self addViewBackedTopBorderWithHeight:height color:color leftOffset:0 rightOffset:0 andTopOffset:0];
}


//////////
// Top + Offset
//////////

-(CALayer*)createTopBorderWithHeight: (CGFloat)height color:(UIColor*)color leftOffset:(CGFloat)leftOffset rightOffset:(CGFloat)rightOffset andTopOffset:(CGFloat)topOffset {
    // Subtract the bottomOffset from the height and the thickness to get our final y position.
    // Add a left offset to our x to get our x position.
    // Minus our rightOffset and negate the leftOffset from the width to get our endpoint for the border.
    return [self getOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, 0 + topOffset, self.frame.size.width - leftOffset - rightOffset, height) andColor:color];
}

-(UIView*)createViewBackedTopBorderWithHeight: (CGFloat)height color:(UIColor*)color leftOffset:(CGFloat)leftOffset rightOffset:(CGFloat)rightOffset andTopOffset:(CGFloat)topOffset {
    UIView *border = [self getViewBackedOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, 0 + topOffset, self.frame.size.width - leftOffset - rightOffset, height) andColor:color];
    border.autoresizingMask = UIViewAutoresizingFlexibleWidth|UIViewAutoresizingFlexibleBottomMargin;
    return border;
}

-(void)addTopBorderWithHeight: (CGFloat)height color:(UIColor*)color leftOffset:(CGFloat)leftOffset rightOffset:(CGFloat)rightOffset andTopOffset:(CGFloat)topOffset {
    // Add leftOffset to our X to get start X position.
    // Add topOffset to Y to get start Y position
    // Subtract left offset from width to negate shifting from leftOffset.
    // Subtract rightoffset from width to set end X and Width.
    [self addOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, 0 + topOffset, self.frame.size.width - leftOffset - rightOffset, height) andColor:color];
}

-(void)addViewBackedTopBorderWithHeight: (CGFloat)height color:(UIColor*)color leftOffset:(CGFloat)leftOffset rightOffset:(CGFloat)rightOffset andTopOffset:(CGFloat)topOffset {
    UIView *border = [self createViewBackedTopBorderWithHeight:height color:color leftOffset:leftOffset rightOffset:rightOffset andTopOffset:topOffset];
    [self addSubview:border];
}


//////////
// Right
//////////

-(CALayer*)createRightBorderWithWidth: (CGFloat)width andColor:(UIColor*)color{
    return [self createRightBorderWithWidth:width color:color rightOffset:0 topOffset:0 andBottomOffset:0];
}

-(UIView*)createViewBackedRightBorderWithWidth: (CGFloat)width andColor:(UIColor*)color{
    return [self createViewBackedRightBorderWithWidth:width color:color rightOffset:0 topOffset:0 andBottomOffset:0];
}

-(void)addRightBorderWithWidth: (CGFloat)width andColor:(UIColor*)color{
    [self addRightBorderWithWidth:width color:color rightOffset:0 topOffset:0 andBottomOffset:0];
}

-(void)addViewBackedRightBorderWithWidth: (CGFloat)width andColor:(UIColor*)color{
    [self addViewBackedRightBorderWithWidth:width color:color rightOffset:0 topOffset:0 andBottomOffset:0];
}



//////////
// Right + Offset
//////////

-(CALayer*)createRightBorderWithWidth: (CGFloat)width color:(UIColor*)color rightOffset:(CGFloat)rightOffset topOffset:(CGFloat)topOffset andBottomOffset:(CGFloat)bottomOffset{
    
    // Subtract bottomOffset from the height to get our end.
    return [self getOneSidedBorderWithFrame:CGRectMake(self.frame.size.width-width-rightOffset, 0 + topOffset, width, self.frame.size.height - topOffset - bottomOffset) andColor:color];
}

-(UIView*)createViewBackedRightBorderWithWidth: (CGFloat)width color:(UIColor*)color rightOffset:(CGFloat)rightOffset topOffset:(CGFloat)topOffset andBottomOffset:(CGFloat)bottomOffset{
    UIView *border = [self getViewBackedOneSidedBorderWithFrame:CGRectMake(self.frame.size.width-width-rightOffset, 0 + topOffset, width, self.frame.size.height - topOffset - bottomOffset) andColor:color];
    border.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin|UIViewAutoresizingFlexibleHeight;
    return border;
}

-(void)addRightBorderWithWidth: (CGFloat)width color:(UIColor*)color rightOffset:(CGFloat)rightOffset topOffset:(CGFloat)topOffset andBottomOffset:(CGFloat)bottomOffset{
    
    // Subtract the rightOffset from our width + thickness to get our final x position.
    // Add topOffset to our y to get our start y position.
    // Subtract topOffset from our height, so our border doesn't extend past teh view.
    // Subtract bottomOffset from the height to get our end.
    [self addOneSidedBorderWithFrame:CGRectMake(self.frame.size.width-width-rightOffset, 0 + topOffset, width, self.frame.size.height - topOffset - bottomOffset) andColor:color];
}

-(void)addViewBackedRightBorderWithWidth: (CGFloat)width color:(UIColor*)color rightOffset:(CGFloat)rightOffset topOffset:(CGFloat)topOffset andBottomOffset:(CGFloat)bottomOffset{
    UIView *border = [self createViewBackedRightBorderWithWidth:width color:color rightOffset:rightOffset topOffset:topOffset andBottomOffset:bottomOffset];
    [self addSubview:border];
}


//////////
// Bottom
//////////

-(CALayer*)createBottomBorderWithHeight: (CGFloat)height andColor:(UIColor*)color{
    return [self createBottomBorderWithHeight:height color:color leftOffset:0 rightOffset:0 andBottomOffset:0];
}

-(UIView*)createViewBackedBottomBorderWithHeight: (CGFloat)height andColor:(UIColor*)color{
    return [self createViewBackedBottomBorderWithHeight:height color:color leftOffset:0 rightOffset:0 andBottomOffset:0];
}

-(void)addBottomBorderWithHeight: (CGFloat)height andColor:(UIColor*)color{
    [self addBottomBorderWithHeight:height color:color leftOffset:0 rightOffset:0 andBottomOffset:0];
}

-(void)addViewBackedBottomBorderWithHeight: (CGFloat)height andColor:(UIColor*)color{
    [self addViewBackedBottomBorderWithHeight:height color:color leftOffset:0 rightOffset:0 andBottomOffset:0];
}


//////////
// Bottom + Offset
//////////

-(CALayer*)createBottomBorderWithHeight: (CGFloat)height color:(UIColor*)color leftOffset:(CGFloat)leftOffset rightOffset:(CGFloat)rightOffset andBottomOffset:(CGFloat)bottomOffset {
    // Subtract the bottomOffset from the height and the thickness to get our final y position.
    // Add a left offset to our x to get our x position.
    // Minus our rightOffset and negate the leftOffset from the width to get our endpoint for the border.
    return [self getOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, self.frame.size.height-height-bottomOffset, self.frame.size.width - leftOffset - rightOffset, height) andColor:color];
}

-(UIView*)createViewBackedBottomBorderWithHeight: (CGFloat)height color:(UIColor*)color leftOffset:(CGFloat)leftOffset rightOffset:(CGFloat)rightOffset andBottomOffset:(CGFloat)bottomOffset{
    UIView *border = [self getViewBackedOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, self.frame.size.height-height-bottomOffset, self.frame.size.width - leftOffset - rightOffset, height) andColor:color];
    border.autoresizingMask = UIViewAutoresizingFlexibleWidth|UIViewAutoresizingFlexibleTopMargin;
    return border;
}

-(void)addBottomBorderWithHeight: (CGFloat)height color:(UIColor*)color leftOffset:(CGFloat)leftOffset rightOffset:(CGFloat)rightOffset andBottomOffset:(CGFloat)bottomOffset {
    // Subtract the bottomOffset from the height and the thickness to get our final y position.
    // Add a left offset to our x to get our x position.
    // Minus our rightOffset and negate the leftOffset from the width to get our endpoint for the border.
    [self addOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, self.frame.size.height-height-bottomOffset, self.frame.size.width - leftOffset - rightOffset, height) andColor:color];
}

-(void)addViewBackedBottomBorderWithHeight: (CGFloat)height color:(UIColor*)color leftOffset:(CGFloat)leftOffset rightOffset:(CGFloat)rightOffset andBottomOffset:(CGFloat)bottomOffset{
    UIView *border = [self createViewBackedBottomBorderWithHeight:height color:color leftOffset:leftOffset rightOffset:rightOffset andBottomOffset:bottomOffset];
    [self addSubview:border];
}



//////////
// Left
//////////

-(CALayer*)createLeftBorderWithWidth: (CGFloat)width andColor:(UIColor*)color{
    return [self createLeftBorderWithWidth:width color:color leftOffset:0 topOffset:0 andBottomOffset:0];
}

-(UIView*)createViewBackedLeftBorderWithWidth: (CGFloat)width andColor:(UIColor*)color{
    return [self createViewBackedLeftBorderWithWidth:width color:color leftOffset:0 topOffset:0 andBottomOffset:0];
}

-(void)addLeftBorderWithWidth: (CGFloat)width andColor:(UIColor*)color{
    [self addLeftBorderWithWidth:width color:color leftOffset:0 topOffset:0 andBottomOffset:0];
}

-(void)addViewBackedLeftBorderWithWidth: (CGFloat)width andColor:(UIColor*)color{
    [self addViewBackedLeftBorderWithWidth:width color:color leftOffset:0 topOffset:0 andBottomOffset:0];
}



//////////
// Left + Offset
//////////

-(CALayer*)createLeftBorderWithWidth: (CGFloat)width color:(UIColor*)color leftOffset:(CGFloat)leftOffset topOffset:(CGFloat)topOffset andBottomOffset:(CGFloat)bottomOffset {
    return [self getOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, 0 + topOffset, width, self.frame.size.height - topOffset - bottomOffset) andColor:color];
}

-(UIView*)createViewBackedLeftBorderWithWidth: (CGFloat)width color:(UIColor*)color leftOffset:(CGFloat)leftOffset topOffset:(CGFloat)topOffset andBottomOffset:(CGFloat)bottomOffset{
    UIView *border = [self getViewBackedOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, 0 + topOffset, width, self.frame.size.height - topOffset - bottomOffset) andColor:color];
    border.autoresizingMask = UIViewAutoresizingFlexibleRightMargin|UIViewAutoresizingFlexibleHeight;
    return border;
}

-(void)addLeftBorderWithWidth: (CGFloat)width color:(UIColor*)color leftOffset:(CGFloat)leftOffset topOffset:(CGFloat)topOffset andBottomOffset:(CGFloat)bottomOffset {
    [self addOneSidedBorderWithFrame:CGRectMake(0 + leftOffset, 0 + topOffset, width, self.frame.size.height - topOffset - bottomOffset) andColor:color];
}

-(void)addViewBackedLeftBorderWithWidth: (CGFloat)width color:(UIColor*)color leftOffset:(CGFloat)leftOffset topOffset:(CGFloat)topOffset andBottomOffset:(CGFloat)bottomOffset{
    UIView *border = [self createViewBackedLeftBorderWithWidth:width color:color leftOffset:leftOffset topOffset:topOffset andBottomOffset:bottomOffset];
    [self addSubview:border];
}



//////////
// Private: Our methods call these to add their borders.
//////////

-(void)addOneSidedBorderWithFrame:(CGRect)frame andColor:(UIColor*)color
{
    CALayer *border = [self getOneSidedBorderWithFrame:frame andColor:color];
    [self.layer addSublayer:border];
}

-(CALayer*)getOneSidedBorderWithFrame:(CGRect)frame andColor:(UIColor*)color
{
    CALayer *border = [CALayer layer];
    border.frame = frame;
    [border setBackgroundColor:color.CGColor];
    return border;
}

-(UIView*)getViewBackedOneSidedBorderWithFrame:(CGRect)frame andColor:(UIColor*)color
{
    UIView *border = [[UIView alloc]initWithFrame:frame];
    [border setBackgroundColor:color];
    return border;
}


@end

File: UIView-Borders/UIView-Borders/UIView+Borders.swift
//
//  UIView+Borders.swift
//  UIView-Borders
//
//  Created by Tommie N. Carter, Jr., MBA on 3/27/17.
//  Copyright © 2017 MING Technology. All rights reserved.
//

import UIKit

extension UIView {
  //////////
  // Top
  //////////
  func createTopBorder(with height:CGFloat, and color: UIColor) -> CALayer {
    return self.createTopBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
  }
  func createViewBackedTopBorder(with height:CGFloat, and color: UIColor) -> UIView {
    return self.createViewBackedTopBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
  }
  func addTopBorder(with height:CGFloat, and color: UIColor) {
    self.addTopBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
  }
  func addViewBackedTopBorder(with height:CGFloat, and color: UIColor)  {
    self.addViewBackedTopBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
  }
  //////////
  // Top + Offset
  //////////
  func createTopBorder(with height: CGFloat, color:UIColor, leftOffset:CGFloat, rightOffset:CGFloat, and topOffset:CGFloat) -> CALayer {
    // Subtract the bottomOffset from the height and the thickness to get our final y position.
    // Add a left offset to our x to get our x position.
    // Minus our rightOffset and negate the leftOffset from the width to get our endpoint for the border.
    return self.getOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: 0 + topOffset, width: self.frame.size.width - leftOffset - rightOffset, height: height), and: color)
  }
  func createViewBackedTopBorder(with height: CGFloat, color:UIColor, leftOffset:CGFloat, rightOffset:CGFloat, and topOffset:CGFloat) ->UIView {
    let borderView = self.getViewBackedOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: 0 + topOffset, width: self.frame.size.width - leftOffset - rightOffset, height: height), and: color)
    borderView.autoresizingMask = [.flexibleWidth,.flexibleBottomMargin] //UIViewAutoresizingFlexibleWidth|UIViewAutoresizingFlexibleBottomMargin
    return borderView
  }
  func addTopBorder(with height: CGFloat, color:UIColor, leftOffset:CGFloat, rightOffset:CGFloat, and topOffset:CGFloat)  {
    // Add leftOffset to our X to get start X position.
    // Add topOffset to Y to get start Y position
    // Subtract left offset from width to negate shifting from leftOffset.
    // Subtract rightoffset from width to set end X and Width.
    self.addOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: 0 + topOffset, width: self.frame.size.width - leftOffset - rightOffset, height: height), and: color)
  }
  func addViewBackedTopBorder(with height: CGFloat, color:UIColor, leftOffset:CGFloat, rightOffset:CGFloat, and topOffset:CGFloat)  {
    let borderView = self.createViewBackedTopBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
    self.addSubview(borderView)
  }
  //////////
  // Right
  //////////
  func createRightBorder(with width:CGFloat, and color: UIColor) -> CALayer {
    return self.createRightBorder(with: width, color: color, rightOffset: 0.0, topOffset: 0.0, and: 0.0)
  }
  func createViewBackedRightBorder(with width:CGFloat, and color: UIColor) -> UIView {
    return self.createViewBackedRightBorder(with: width, color: color, rightOffset: 0.0, topOffset: 0.0, and: 0.0)
  }
  func addRightBorder(with width:CGFloat, and color: UIColor) {
    self.addRightBorder(with: width, color: color, rightOffset: 0.0, topOffset: 0.0, and: 0.0)
  }
  func addViewBackedRightBorder(with width:CGFloat, and color: UIColor)  {
    self.addViewBackedRightBorder(with: width, color: color, rightOffset: 0.0, topOffset: 0.0, and: 0.0)
  }
  //////////
  // Right + Offset
  //////////
  func createRightBorder( with width: CGFloat, color:UIColor, rightOffset:CGFloat, topOffset:CGFloat, and bottomOffset:CGFloat) ->CALayer {
    // Subtract bottomOffset from the height to get our end.
    return getOneSidedBorder(with: CGRect(x: self.frame.size.width-width-rightOffset, y: 0 + topOffset, width: width, height: self.frame.size.height - topOffset - bottomOffset), and: color)
  }
   func createViewBackedRightBorder(with width: CGFloat, color:UIColor, rightOffset:CGFloat, topOffset:CGFloat, and bottomOffset:CGFloat) -> UIView {
    let borderView = self.getViewBackedOneSidedBorder(with: CGRect(x: self.frame.size.width-width-rightOffset, y: 0 + topOffset, width: width, height: self.frame.size.height - topOffset - bottomOffset) , and: color)
    borderView.autoresizingMask = [.flexibleLeftMargin, .flexibleHeight]//UIViewAutoresizingFlexibleLeftMargin|UIViewAutoresizingFlexibleHeight
    return borderView
  }
  func addRightBorder( with width: CGFloat, color:UIColor, rightOffset:CGFloat, topOffset:CGFloat, and bottomOffset:CGFloat)  {
  
    // Subtract the rightOffset from our width + thickness to get our final x position.
    // Add topOffset to our y to get our start y position.
    // Subtract topOffset from our height, so our border doesn't extend past teh view.
    // Subtract bottomOffset from the height to get our end.
    self.addOneSidedBorder(with: CGRect(x: self.frame.size.width-width-rightOffset, y: 0 + topOffset, width: width, height: self.frame.size.height - topOffset - bottomOffset) , and: color)
  }
  func addViewBackedRightBorder(with width: CGFloat, color:UIColor, rightOffset:CGFloat, topOffset:CGFloat, and bottomOffset:CGFloat)  {
    let borderView = self.createViewBackedRightBorder(with: width, color: color, rightOffset: 0.0, topOffset: 0.0, and: 0.0)
    self.addSubview(borderView)
  }
  //////////
  // Bottom
  //////////
  func createBottomBorder(with height:CGFloat, and color: UIColor) -> CALayer {
    return self.createBottomBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
  }
  func createViewBackedBottomBorder(with height:CGFloat, and color: UIColor) -> UIView {
    return self.createViewBackedBottomBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
  }
  func addBottomBorder(with height:CGFloat, and color: UIColor) {
    self.addBottomBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
  }
  func addViewBackedBottomBorder(with height:CGFloat, and color: UIColor)  {
    self.addViewBackedBottomBorder(with: height, color: color, leftOffset: 0.0, rightOffset: 0.0, and: 0.0)
  }
  //////////
  // Bottom + Offset
  //////////
  func createBottomBorder(with height: CGFloat, color:UIColor, leftOffset:CGFloat, rightOffset:CGFloat, and bottomOffset:CGFloat) -> CALayer {
    // Subtract the bottomOffset from the height and the thickness to get our final y position.
    // Add a left offset to our x to get our x position.
    // Minus our rightOffset and negate the leftOffset from the width to get our endpoint for the border.
    return self.getOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: self.frame.size.height-height-bottomOffset, width: self.frame.size.width - leftOffset - rightOffset, height: height), and: color)
  }
  func createViewBackedBottomBorder(with height: CGFloat, color:UIColor, leftOffset:CGFloat, rightOffset:CGFloat, and bottomOffset:CGFloat) ->UIView {
    let borderView = self.getViewBackedOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: self.frame.size.height-height-bottomOffset, width: self.frame.size.width - leftOffset - rightOffset, height: height), and: color)
    borderView.autoresizingMask = [.flexibleWidth, .flexibleTopMargin] //UIViewAutoresizingFlexibleWidth|UIViewAutoresizingFlexibleTopMargin;
  
    return borderView
  }
  func addBottomBorder(with height: CGFloat, color:UIColor, leftOffset:CGFloat, rightOffset:CGFloat, and bottomOffset:CGFloat)  {
    // Subtract the bottomOffset from the height and the thickness to get our final y position.
    // Add a left offset to our x to get our x position.
    // Minus our rightOffset and negate the leftOffset from the width to get our endpoint for the border.
    self.addOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: self.frame.size.height-height-bottomOffset, width: self.frame.size.width - leftOffset - rightOffset,height: height), and: color)
  }
  func addViewBackedBottomBorder(with height: CGFloat, color:UIColor, leftOffset:CGFloat, rightOffset:CGFloat, and bottomOffset:CGFloat)  {
    let borderView = self.createViewBackedBottomBorder(with: height, color: color, leftOffset: leftOffset, rightOffset: rightOffset, and: bottomOffset)
    self.addSubview(borderView)
  }
  //////////
  // Left
  //////////
  func createLeftBorder(with width:CGFloat, and color: UIColor) -> CALayer {
    return self.createLeftBorder(with: width, color: color, leftOffset: 0.0, topOffset: 0.0, and: 0.0)
  
  }
  func createViewBackedLeftBorder(with width:CGFloat, and color: UIColor) -> UIView {
    return createViewBackedLeftBorder(with: width, color: color, leftOffset: 0.0, topOffset: 0.0, and: 0.0)
  }
  func addLeftBorder(with width:CGFloat, and color: UIColor) {
    self.addLeftBorder(with: width, color: color, leftOffset: 0.0, topOffset: 0.0, and: 0.0)
  }
  func addViewBackedLeftBorder(with width:CGFloat, and color: UIColor)  {
    
    self.addViewBackedLeftBorder(with: width, color: color, leftOffset: 0.0, topOffset: 0.0, and: 0.0)
    
  }
  //////////
  // Left + Offset
  //////////
  func createLeftBorder( with width: CGFloat, color:UIColor, leftOffset:CGFloat, topOffset:CGFloat, and bottomOffset:CGFloat) -> CALayer {
  
    return getOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: 0 + topOffset, width: width, height: self.frame.size.height - topOffset - bottomOffset) , and: color)
  }
  func createViewBackedLeftBorder(with width: CGFloat, color:UIColor, leftOffset:CGFloat, topOffset:CGFloat, and bottomOffset:CGFloat) -> UIView {
    let borderView = getViewBackedOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: 0 + topOffset, width: width, height: self.frame.size.height - topOffset - bottomOffset) , and: color)
    borderView.autoresizingMask = [.flexibleRightMargin , .flexibleHeight] //UIViewAutoresizingFlexibleRightMargin|UIViewAutoresizingFlexibleHeight;
    return borderView
  }
  func addLeftBorder( with width: CGFloat, color:UIColor, leftOffset:CGFloat, topOffset:CGFloat, and bottomOffset:CGFloat)  {
    self.addOneSidedBorder(with: CGRect(x: 0 + leftOffset, y: 0 + topOffset, width: width, height: self.frame.size.height - topOffset - bottomOffset), and: color)
  }
  func addViewBackedLeftBorder(with width: CGFloat, color:UIColor, leftOffset:CGFloat, topOffset:CGFloat, and bottomOffset:CGFloat)  {
    let borderView = self.createViewBackedLeftBorder(with: width, color: color, leftOffset: leftOffset, topOffset: topOffset, and: bottomOffset)
    self.addSubview(borderView)
  }
  //////////
  // Private: Our methods call these to add their borders.
  //////////
  func addOneSidedBorder(with frame:CGRect, and color:UIColor){
    let borderLayer = self.getOneSidedBorder(with: frame, and: color)
    self.layer.addSublayer(borderLayer)
  }
  func getOneSidedBorder(with frame:CGRect, and color:UIColor) -> CALayer {
    let borderLayer = CALayer()
    borderLayer.frame = frame
    borderLayer.backgroundColor = color.cgColor
    return borderLayer
  
  }
  func getViewBackedOneSidedBorder(with frame:CGRect, and color:UIColor) -> UIView {
    let borderView = UIView(frame: frame)
    borderView.backgroundColor = color
    return borderView
  }
}
